backend/service/topology/querybuilder.go

```diff
@@ -0,0 +1,115 @@
+package topology
+
+import (
+	"fmt"
+	"strings"
+
+	sq "github.com/Masterminds/squirrel"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+const (
+	column   = "column."
+	metadata = "metadata."
+)
+
+func paginatedQueryBuilder(
+	filter *topologyv1.SearchTopologyRequest_Filter,
+	sort *topologyv1.SearchTopologyRequest_Sort,
+	pageToken,
+	limit int,
+) (sq.SelectBuilder, error) {
+	queryLimit := 100
+	if limit > 0 {
+		queryLimit = limit
+	}
+
+	queryOffset := 0
+	if pageToken > 0 {
+		queryOffset = pageToken * limit
+	}
+
+	query := sq.StatementBuilder.PlaceholderFormat(sq.Dollar).
+		Select("id, data, metadata").
+		From("topology_cache").
+		Limit(uint64(queryLimit)).
+		Offset(uint64(queryOffset))
+
+	query = filterQueryBuilder(query, filter)
+	query = sortQueryBuilder(query, sort)
+
+	return query, nil
+}
+
+func filterQueryBuilder(query sq.SelectBuilder, f *topologyv1.SearchTopologyRequest_Filter) sq.SelectBuilder {
+	if f.Search != nil && len(f.Search.Field) > 0 {
+		if strings.HasPrefix(f.Search.Field, column) {
+			query = query.Where(sq.Like{strings.TrimPrefix(f.Search.Field, column): fmt.Sprintf("%%%s%%", f.Search.Text)})
```

Comment:
i'm fairly certain you cannot trust user input on the left side of the these clauses without the risk of injection. only the right hand side is converted to an argument which is safe from injection.

Reply:
Let me do some more testing and add some sql injection tests as well.

Reply:
did some poking around and i think for JSONB columns you can use the contains operator to get around this https://stackoverflow.com/a/62895355. though on second look i don't think that will work the LIKE % syntax.

different use case and methodology https://stackoverflow.com/a/45849743 but i think it's probably going to look something like that using the JSON path language and passing a expression built from user input as an argument.

that led me to some really good details here: https://blog.sql-workbench.eu/post/json-path/, especially under `More complex queries`. it does mention some impact on the query plan and perf implications but i'm not sure it would matter even with row counts on the order of thousands where the entire set probably fits in the DB's memory.

maybe even another option is to only allow alphanumeric values as left-hand side input, which i think are impossible to perform injection attacks with. or some other form of escaping. more research and tinkering required.

Reply:
Nice some of the metadata queries would be much easier to construct with some of this, I'm going to incorporate some of this. 

Ive also found a way to parameterize all of these user inputs with squirrel by using [Expressions](https://godoc.org/github.com/Masterminds/squirrel#Expr).